databricks_mcp/src/databricks_mcp/connector.py

```diff
@@ -0,0 +1,572 @@
+"""MCP OAuth Discovery and Dynamic Client Registration for Databricks.
+
+This module provides functionality for discovering OAuth metadata and performing
+Dynamic Client Registration (DCR) for MCP (Model Context Protocol) connections.
+"""
+
+import json
+import logging
+import re
+from typing import Any, Dict, Optional, Tuple
+from urllib.parse import urljoin, urlparse
+
+import requests
+from databricks.sdk import WorkspaceClient
+from databricks.sdk.service import catalog
+from databricks_ai_bridge.utils.annotations import experimental
+
+# Configure logging
+logger = logging.getLogger(__name__)
+
+# Constants
+DEFAULT_TIMEOUT = 10
+DEFAULT_PORT = "443"
+WWW_AUTHENTICATE_HEADER = "WWW-Authenticate"
+CONTENT_TYPE_JSON = "application/json"
+
+# ---------------- Helpers ----------------
+
+
+def _fetch_json(url: str, description: str) -> Dict[str, Any]:
+    """Fetch and parse JSON from a URL.
+
+    Args:
+        url: The URL to fetch from
+        description: Human-readable description for error messages
+
+    Returns:
+        Parsed JSON as a dictionary
+
+    Raises:
+        RuntimeError: If the request fails or JSON parsing fails
+    """
+    logger.debug(f"Fetching {description} from {url}")
+
+    try:
+        resp = requests.get(url, timeout=DEFAULT_TIMEOUT)
+        resp.raise_for_status()
+    except requests.exceptions.Timeout as e:
+        raise RuntimeError(f"Timeout while fetching {description} from {url}") from e
+    except requests.exceptions.RequestException as e:
+        raise RuntimeError(f"Failed to fetch {description} from {url}: {e}") from e
+
+    try:
+        data = resp.json()
+        logger.debug(f"Successfully fetched {description}")
+        return data
+    except ValueError as e:
+        raise RuntimeError(
+            f"Failed to parse {description} from {url} as JSON: {e}. Response: {resp.text[:200]}"
+        ) from e
+
+
+def _parse_www_authenticate(header_value: str) -> Dict[str, str]:
+    """Parse WWW-Authenticate header value into key-value parameters.
+
+    Args:
+        header_value: The WWW-Authenticate header value
+
+    Returns:
+        Dictionary of parsed parameters from the header
+    """
+    if not header_value:
+        return {}
+
+    parts = header_value.split(" ", 1)
+    params_part = parts[1] if len(parts) == 2 else parts[0]
+    params = {}
+
+    for m in re.finditer(r'(\w+)\s*=\s*"([^"]*)"', params_part):
+        params[m.group(1)] = m.group(2)
```

Comment:
Looks like we are capturing the `"` marks in the regex above, can we remove them when adding to the dictionary?